platform/darwin/src/MGLMapSnapshotter.mm

```diff
@@ -0,0 +1,163 @@
+#import "MGLGeometry_Private.h"
+#import "NSBundle+MGLAdditions.h"
+
+#if TARGET_OS_IPHONE
+#import "UIImage+MGLAdditions.h"
+#else
+#import "NSImage+MGLAdditions.h"
+#endif
+
+@implementation MGLMapSnapshotOptions
+
+- (instancetype _Nonnull)initWithStyleURL:(NSURL* _Nonnull)styleURL camera:(MGLMapCamera*)camera size:(CGSize) size;
+{
+    self = [super init];
+    if (self) {
+        _styleURL = styleURL;
+        _size = size;
+        _camera = camera;
+#if TARGET_OS_IPHONE
+        _scale = [UIScreen mainScreen].scale;
+#else
+        _scale = [NSScreen mainScreen].backingScaleFactor;
+#endif
+        
+    }
+    return self;
+}
+
+@end
+
+@implementation MGLMapSnapshotter {
+    
+    std::shared_ptr<mbgl::ThreadPool> mbglThreadPool;
+    std::unique_ptr<mbgl::MapSnapshotter> mbglMapSnapshotter;
+    std::unique_ptr<mbgl::Actor<mbgl::MapSnapshotter::Callback>> snapshotCallback;
+}
+
+- (instancetype)initWithOptions:(MGLMapSnapshotOptions*)options;
+{
+    self = [super init];
+    if (self) {
+        _loading = false;
+        
+        mbgl::DefaultFileSource *mbglFileSource = [MGLOfflineStorage sharedOfflineStorage].mbglFileSource;
+        mbglThreadPool = mbgl::sharedThreadPool();
+        
+        std::string styleURL = std::string([options.styleURL.absoluteString UTF8String]);
+        
+        // Size; taking into account the minimum texture size for OpenGL ES
+        mbgl::Size size = {
+            static_cast<uint32_t>(MAX(options.size.width, 64)),
+            static_cast<uint32_t>(MAX(options.size.height, 64))
+        };
+        
+        float pixelRatio = MAX(options.scale, 1);
+        
+        // Camera options
+        mbgl::CameraOptions cameraOptions;
+        if (CLLocationCoordinate2DIsValid(options.camera.centerCoordinate)) {
+            cameraOptions.center = MGLLatLngFromLocationCoordinate2D(options.camera.centerCoordinate);
+        }
+        cameraOptions.angle = MAX(0, options.camera.heading) * mbgl::util::DEG2RAD;
+        cameraOptions.zoom = MAX(0, options.zoom);
+        cameraOptions.pitch = MAX(0, options.camera.pitch);
+        
+        // Region
+        mbgl::optional<mbgl::LatLngBounds> region;
+        if (!MGLCoordinateBoundsIsEmpty(options.region)) {
+            region = MGLLatLngBoundsFromCoordinateBounds(options.region);
+        }
+        
+        // Create the snapshotter
+        mbglMapSnapshotter = std::make_unique<mbgl::MapSnapshotter>(*mbglFileSource, *mbglThreadPool, styleURL, size, pixelRatio, cameraOptions, region);
+    }
+    return self;
+}
+
+- (void)startWithCompletionHandler:(MGLMapSnapshotCompletionHandler)completion;
+{
+    [self startWithQueue:dispatch_get_main_queue() completionHandler:completion];
+}
+
+- (void)startWithQueue:(dispatch_queue_t)queue completionHandler:(MGLMapSnapshotCompletionHandler)completion;
+{
+    if ([self isLoading]) {
+        NSDictionary *userInfo = @{NSLocalizedDescriptionKey: @"Already started this snapshotter"};
```

Comment:
This string should be [localizable](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/ios/DEVELOPING.md#adding-user-facing-text), since it can be presented to the user.